Final Project - Cryptoanalyser

Viktor7783 opened the pull request:
В рамках данного проекта я написал программу для шифрования и расшифровки текста, используя шифр Цезаря.
Программа запускается с класса Main.

Предлагается 4 вида действий:
1 - Шифровка текста из заданного файла
2 - Расшифровка при известном ключе
3 - Расшифровка методом brute force
4 - Расшифровка методом статистического анализа

Я решил объединить два алфавита Русский и Английский в один List: думаю в тексте встречаются одновременно и русские и английские слова, которые нужно зашифровать, чтобы защитить информацию.

Я не добавлял знаки пунктуации и пробел в свой List с алфавитом, так как считаю, что если зашифровать например пробел (допустим он стал буквой 'п') - то его очень легко выявить среди зашифрованного текста: 
он будет повторяться очень много раз, а если в текст добавить слова из другого алфавита (не русского и английского) - пробел станет еще заметнее... Останется лишь перебрать все возможные ключи всего лишь для одной буквы 'п', чтобы она вновь стала пробелом и с данным ключом расшифровать весь текст.

Первые два пункта были для меня предельно понятны (Досконально разобраны на сайте).
Последние два дались туго, чтобы понять, что от меня хотят - основное время убил на это:)
Очень помогла лекция Факультатив -1!!!

Для пункта 3 и 4 ввёл ограничение: файл должен быть не менее 1000 символов.

Расшифровку методом brute force я реализовал:
 1. Выявлял в файле-примере 3 наиболее часто встречающихся в нём слов.
 2.  Перекодировал ВЕСЬ ТЕКСТ из зашифрованного файла с ключём с шагом в единицу.
 3. Выявлял для каждого ключа во всём зашифрованном тексте 3 часто встречающихся слова и сравнивал их с 3-мя словами, добытыми из файла-примера.
 4. Если 3 частых слова из файла-примера совпадали с 3-мя частыми словами из шифрованного текста - записывал в файл текст, щифрованный ключом на этом шаге.
 
Расшифровку методом статистического анализа я реализовал: (для себя я понял это задание так...)
1. Выявлял в файле-примере 3 наиболее часто встречающихся в нём слов. (list1) 
2. Выявлял в зашифрованном файле 3 наиболее часто встречающихся в нём слов. (list2)
3. Перекодировал в list2 всего лишь ТРИ СЛОВА с помощью ключа с шагом 1, до тех пор, пока они не станут эквивалентны словам в list1. Сохранял данный ключ - key.
4. Расшифровывал теперь уже весь файл с данным ключом key и записывал его в файл.

Спасибо, что дочитал:)

Viktor7783 committed:
Create class Constants

Viktor7783 committed:
Написал класс FileService.
Написал класс SelectTaskService
Создал Exceptions:
SmallFileException
WrongFileException
WrongNumberException

Viktor7783 committed:
Написал классы:
ReadPathService
CryptoModel
CryptoService
Constants

Изменил классы:
FileService
SelectTaskService

Viktor7783 committed:
Написал класс FileService.
Написал класс SelectTaskService
Создал Exceptions:
SmallFileException
WrongFileException
WrongNumberException

Добавил много нового:)
Изменил почти каждый класс

Viktor7783 committed:
Написал класс FileService.
Написал класс SelectTaskService
Создал Exceptions:
SmallFileException
WrongFileException
WrongNumberException

Добавил много нового:)
Изменил почти каждый класс

Изменил класс CryptoModel:
Добавил последний метод поиска ключа статистическим анализом.
Переделал метод bruteforce.

Viktor7783 committed:
Написал класс FileService.
Написал класс SelectTaskService
Создал Exceptions:
SmallFileException
WrongFileException
WrongNumberException

Добавил много нового:)
Изменил почти каждый класс

Изменил класс CryptoModel:
Добавил последний метод поиска ключа статистическим анализом.
Переделал метод bruteforce.
Добавил приветствие:)

Viktor7783 committed:
Update README.md

antonkupreychik commented on src/services/CryptoModel.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    public void encrypt(String source, String destination, int key) {
+        List<String> sourceList = fileService.readFromFile(source, false);
+        List<String> cryptoList = cryptoCore(sourceList, key, ALPHABET_RUS_ENG);
```
Ты в начале строки уже написал List и нет необходимости еще раз писать list в названии переменной, то есть List<String> sources = fileService....

Viktor7783 replied:
Понял, исправлю, я просто для себя напоминание сделал, что всё, чо я  делаю дальше - делаю с коллекцией List :)))

antonkupreychik commented on src/services/CryptoModel.java:
```diff
@@ -0,0 +1,140 @@
+        }
+    }
+
+    private static List<String> getSourceMaxValueList(List<String> sourceList) {
```
думаю тут можно без статики

Viktor7783 replied:
Да, я сам не знаю откуда здесь взялась статика. Вроде так досоконально проверял весь проект...

antonkupreychik commented on src/services/CryptoModel.java:
```diff
@@ -0,0 +1,140 @@
+                }
+                if (isAlphabetWord) {
+                    Integer valueMap = sourceWordsMap.get(word.toLowerCase());
+                    sourceWordsMap.put(word.toLowerCase(), valueMap == null ? 1 : valueMap + 1);
```
1 в константы как increment_value

antonkupreychik commented on src/services/CryptoModel.java:
```diff
@@ -0,0 +1,140 @@
+
+        Comparator<Map.Entry<String, Integer>> comparator = Map.Entry.comparingByValue();
+        Set<Map.Entry<String, Integer>> set = new HashSet<>(sourceWordsMap.entrySet());
+        for (int i = 0; i < 3; i++) {
```
3 тоже в константы, да в ридми кажется о ней была речь, но все равно надо в константу

antonkupreychik commented on src/services/CryptoService.java:
```diff
@@ -0,0 +1,51 @@
+            cryptoModel.statisticalDecrypt(readPathAndKeyService.getSource(), readPathAndKeyService.getExample(), readPathAndKeyService.getDestination());
+        }
+
+        try {
+            reader.close();
+        } catch (IOException e) {
+            System.out.println(INPUT_ERROR);
+            System.exit(0);
+        }
```
try-with-resources? а вообще зачем это тут?

Viktor7783 replied:
Да, спасибо большое за подсказку, не подумал о try-with-resources!
Вместо Scanner я использовал BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))  из за возникшей проблемы:

Изначально использовал Scanner reader!
Когда я передал свой reader в  selectTaskService = new SelectTaskService(reader) - всё ОК
**Но когда дальше я передал reader в readPathAndKeyService = new ReadPathAndKeyService(reader) -начались проблемы!!!** 
Выводилось в консоль сообщение  - "Введите адрес файла c оригинальным текстом, который нужно зашифровать:" **и сразу выводилось сообщение** "Введите адрес файла в который нужно записать зашифрованный текст:" !!! 
В итоге я не мог ввести адрес файла источника в консоль и сразу перебрасывало на ввод адреса файла - куда нужно записывать файл. 
Я пробовал менять строки местами (reader.nextLine()), закомменитровал один вызов  reader.nextLine(), но всё оказалось бесполезным, почему то всегда пропускается один вызов reader.nextLine()!!! и переходит к следующему вызову reader.nextLine()... Пробовал по разному разместить в коде reader.nextLine(), но всё было тщетно, всегда пропускал один вызов reader.nextLine()....
Почитав о Scanner  я узнал что он потоконебезопасен (не знаю почему у меня он глючит, выполняется же всё в одном потоке main...) - и я решил использовать BufferedReader - с ним всё работает без сбоев.

Viktor7783 replied:
Антон, подскажи пожалуйста:
попробовал сделать try-with-resources для моего BufferedReader reader, но так как инициализация самого reader и классов его использующих происходит в конструкторе public CryptoService()  - то при try-with-resources мой reader автоматически закрывается после выполнения конструктора и в методы 
selectTaskService = new SelectTaskService(reader);
readPathAndKeyService = new ReadPathAndKeyService(reader);
**передаётся  reader который уже нельзя использовать - и  программа завершается с exception...**

Поэтому я и оставил try {  - в конце метода start()
            **reader.close(); - поток же нужо закрыть после использования (но это не точно:))**
        } catch (IOException e) {
            System.out.println(INPUT_ERROR);
            System.exit(0);
        }

Либо необходимо переносить инициализацию всех переменных из конструктора в метод start() класса CryptoService.

Надеюсь я тебя не сильно нагрузил:)
Опять столько букв понаписал!

antonkupreychik commented on src/services/FileValidationService.java:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    public void SmallFileTest(Path path, List<String> stringList) throws SmallFileException {
```
название метода с маленькой буквы

antonkupreychik commented on src/services/FileValidationService.java:
```diff
@@ -0,0 +1,74 @@
+public class FileValidationService {
+
+    public void txtFileTest(Path path) {
+        if (!path.getFileName().toString().endsWith(".txt")) {
```
.txt в константы

antonkupreychik reviewed (approved):
огонь, молодец!

Viktor7783 commented:
Привет, Антон!
Спасибо тебе огромное за то, что нашёл время посмотреть мой запоздалый
проект!
Мне очень приятно, что я огненный молодец!☺️
У меня остался один вопрос по BufferedReader вместо Scanner и try- with-
resource.
Я в проекте написал комментарий-вопрос на твою рекоммендацию, пожалуйста,
если будет возможность посмотри пожалуйста, без тебя не разберусь.

вт, 12 мар. 2024 г. в 00:02, antonkupreychik ***@***.***>:

> ***@***.**** approved this pull request.
>
> огонь, молодец!
> ------------------------------
>
> In src/services/CryptoModel.java
> <https://github.com/Viktor7783/my_cryptoanalyser/pull/1#discussion_r1520424459>
> :
>
> > +        List<String> sourceList = fileService.readFromFile(source, false);
> +        List<String> cryptoList = cryptoCore(sourceList, key, ALPHABET_RUS_ENG);
>
> Ты в начале строки уже написал List и нет необходимости еще раз писать
> list в названии переменной, то есть List sources = fileService....
> ------------------------------
>
> In src/services/CryptoModel.java
> <https://github.com/Viktor7783/my_cryptoanalyser/pull/1#discussion_r1520425099>
> :
>
> > +                isDecrypt = true;
> +                break;
> +            }
> +        }
> +        if (isDecrypt) fileService.writeToFile(destination, cryptoCore(sourceList, key, ALPHABET_RUS_ENG_REVERSE));
> +        else {
> +            try {
> +                throw new DecryptionException(String.format(DECRYPT_ERROR_MESSAGE, Path.of(source).getFileName()));
> +            } catch (DecryptionException e) {
> +                System.out.println(e.getMessage());
> +                System.exit(0);
> +            }
> +        }
> +    }
> +
> +    private static List<String> getSourceMaxValueList(List<String> sourceList) {
>
> думаю тут можно без статики
> ------------------------------
>
> In src/services/CryptoModel.java
> <https://github.com/Viktor7783/my_cryptoanalyser/pull/1#discussion_r1520425490>
> :
>
> > +        Map<String, Integer> sourceWordsMap = new HashMap<>();
> +        List<String> sourceMaxValueList = new ArrayList<>();
> +
> +        for (String textLine : sourceList) {
> +            String[] wordsArray = textLine.split(" ");
> +            for (String word : wordsArray) {
> +                boolean isAlphabetWord = true;
> +                for (char c : word.toCharArray()) {
> +                    if (!ALPHABET_RUS_ENG_REVERSE.contains(c)) {
> +                        isAlphabetWord = false;
> +                        break;
> +                    }
> +                }
> +                if (isAlphabetWord) {
> +                    Integer valueMap = sourceWordsMap.get(word.toLowerCase());
> +                    sourceWordsMap.put(word.toLowerCase(), valueMap == null ? 1 : valueMap + 1);
>
> 1 в константы как increment_value
> ------------------------------
>
> In src/services/CryptoModel.java
> <https://github.com/Viktor7783/my_cryptoanalyser/pull/1#discussion_r1520426251>
> :
>
> > +                    if (!ALPHABET_RUS_ENG_REVERSE.contains(c)) {
> +                        isAlphabetWord = false;
> +                        break;
> +                    }
> +                }
> +                if (isAlphabetWord) {
> +                    Integer valueMap = sourceWordsMap.get(word.toLowerCase());
> +                    sourceWordsMap.put(word.toLowerCase(), valueMap == null ? 1 : valueMap + 1);
> +                }
> +
> +            }
> +        }
> +
> +        Comparator<Map.Entry<String, Integer>> comparator = Map.Entry.comparingByValue();
> +        Set<Map.Entry<String, Integer>> set = new HashSet<>(sourceWordsMap.entrySet());
> +        for (int i = 0; i < 3; i++) {
>
> 3 тоже в константы, да в ридми кажется о ней была речь, но все равно надо
> в константу
> ------------------------------
>
> In src/services/CryptoService.java
> <https://github.com/Viktor7783/my_cryptoanalyser/pull/1#discussion_r1520426933>
> :
>
> > +        try {
> +            reader.close();
> +        } catch (IOException e) {
> +            System.out.println(INPUT_ERROR);
> +            System.exit(0);
> +        }
>
> try-with-resources? а вообще зачем это тут?
> ------------------------------
>
> In src/services/FileValidationService.java
> <https://github.com/Viktor7783/my_cryptoanalyser/pull/1#discussion_r1520427320>
> :
>
> > +            }
> +        }
> +    }
> +
> +    public void fileExistTest(Path path) {
> +        if (Files.notExists(path)) {
> +            try {
> +                throw new WrongFileException(String.format(FILE_NOT_EXIST_ERROR, path.getFileName()));
> +            } catch (WrongFileException e) {
> +                System.out.println(e.getMessage());
> +                System.exit(0);
> +            }
> +        }
> +    }
> +
> +    public void SmallFileTest(Path path, List<String> stringList) throws SmallFileException {
>
> название метода с маленькой буквы
> ------------------------------
>
> In src/services/FileValidationService.java
> <https://github.com/Viktor7783/my_cryptoanalyser/pull/1#discussion_r1520428134>
> :
>
> > @@ -0,0 +1,74 @@
> +package services;
> +
> +import exceptions.SmallFileException;
> +import exceptions.WrongFileException;
> +
> +import java.nio.file.Files;
> +import java.nio.file.Path;
> +import java.util.List;
> +
> +import static constants.Constants.*;
> +
> +public class FileValidationService {
> +
> +    public void txtFileTest(Path path) {
> +        if (!path.getFileName().toString().endsWith(".txt")) {
>
> .txt в константы
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Viktor7783/my_cryptoanalyser/pull/1#pullrequestreview-1929276172>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AX6EITHC7R5PYAXHWFOYDBTYXYLV3AVCNFSM6AAAAABEFMX3NOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMRZGI3TMMJXGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Viktor7783 committed:
Написал класс FileService.
Написал класс SelectTaskService
Создал Exceptions:
SmallFileException
WrongFileException
WrongNumberException

Добавил много нового:)
Изменил почти каждый класс

Изменил класс CryptoModel:
Добавил последний метод поиска ключа статистическим анализом.
Переделал метод bruteforce.
Добавил приветствие:)
